src/OmniSharp.Cake/CachingScriptMetadataResolver.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace OmniSharp.Cake
+{
+    internal class CachingScriptMetadataResolver : MetadataReferenceResolver
+    {
+        private static readonly Dictionary<string, ImmutableArray<PortableExecutableReference>> DirectReferenceCache = new Dictionary<string, ImmutableArray<PortableExecutableReference>>();
+        private static readonly Dictionary<string, PortableExecutableReference> MissingReferenceCache = new Dictionary<string, PortableExecutableReference>();
+        private static readonly MetadataReferenceResolver DefaultRuntimeResolver = ScriptMetadataResolver.Default;
+
+        public override bool Equals(object other)
+        {
+            return DefaultRuntimeResolver.Equals(other);
+        }
+
+        public override int GetHashCode()
+        {
+            return DefaultRuntimeResolver.GetHashCode();
+        }
+
+        public override bool ResolveMissingAssemblies => DefaultRuntimeResolver.ResolveMissingAssemblies;
+
+        public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
+        {
+            if (MissingReferenceCache.ContainsKey(referenceIdentity.Name))
+            {
+                return MissingReferenceCache[referenceIdentity.Name];
+            }
```

Comment:
minor nit: Could this use `TryGetValue` rather than doing two dictionary look ups?